src/Features/LanguageServer/Protocol/Extensions/Extensions.cs

```diff
@@ -30,6 +29,19 @@ public static Uri GetURI(this TextDocument document)
                 : ProtocolConversions.GetUriFromFilePath(document.FilePath);
         }
 
+        public static Uri GetUriFromName(this TextDocument document)
+        {
+            Contract.ThrowIfNull(document.FilePath);
+            Contract.ThrowIfNull(document.Name);
+
+            var directoryName = Path.GetDirectoryName(document.FilePath);
+            Contract.ThrowIfNull(directoryName);
+            var path = Path.Combine(directoryName, document.Name);
+            return document is SourceGeneratedDocument
```

Comment:
When call `Document.WithName(string)`   https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/Document.cs,342
It will finally create a new Document Attribute with new name
https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/DocumentState.cs,405

The file path, document id and other stuff in [DocumentInfo](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/DocumentInfo.cs,150) would stay the same

So in order to get the Url with the new name, here I have to parse the directory path and combine it with the name

Reply:
And there is also a field called document folders in document attributes, https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/DocumentInfo.cs,167

Should I respect this as the containing directory, or should I respect `document.FilePath`?

Might be a good question for @jasonmalinowski

Reply:
Gotcha.  So the renamed document doesn't have an updated file path.  Yeah I'm not sure if you should use file path or document folders... what is triggering the rename?

Reply:
AbstractMoveTypeService, when the type name doesn't match the file name. Rename the file to the type name

Reply:
The "Folders" collection represents the virtual folders a document might be present in if you have linked files happening -- a file from a different location on disk can be linked into a folder as it appears in Solution explorer, but that has nothing to do with the actual physical file on disk.

> or call Path.GetDirectoryName on [filepath](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/DocumentInfo.cs,48)?

That.